server/grpc_service.go

```diff
@@ -234,67 +234,91 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
 
-	cluster := s.GetRaftCluster()
-	if cluster == nil {
-		return &pdpb.RegionHeartbeatResponse{Header: s.notBootstrappedHeader()}, nil
-	}
+		if err := s.validateRequest(request.GetHeader()); err != nil {
+			// TODO: How to close this stream?
+			return errors.Trace(err)
+		}
 
-	resp := &pdpb.RegionHeartbeatResponse{}
-	region := newRegionInfo(request.GetRegion(), request.GetLeader())
-	region.DownPeers = request.GetDownPeers()
-	region.PendingPeers = request.GetPendingPeers()
-	region.WrittenBytes = request.GetBytesWritten()
-	if region.GetId() == 0 {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: fmt.Sprintf("invalid request region, %v", request),
+		cluster := s.GetRaftCluster()
+		if cluster == nil {
+			msg := "cluster is not bootstrapped"
+			err = sendErrorRegionHeartbeatResponse(server, s.clusterID, pdpb.ErrorType_NOT_BOOTSTRAPPED, msg)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
 		}
-		resp.Header = s.errorHeader(pberr)
 
-		return resp, nil
-	}
-	if region.Leader == nil {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: fmt.Sprintf("invalid request leader, %v", request),
+		region := newRegionInfo(request.GetRegion(), request.GetLeader())
+		region.DownPeers = request.GetDownPeers()
+		region.PendingPeers = request.GetPendingPeers()
+		region.WrittenBytes = request.GetBytesWritten()
+		if region.GetId() == 0 {
+			msg := fmt.Sprintf("invalid request region, %v", request)
+			err = sendErrorRegionHeartbeatResponse(server, s.clusterID, pdpb.ErrorType_UNKNOWN, msg)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		if region.Leader == nil {
+			msg := fmt.Sprintf("invalid request leader, %v", request)
+			err = sendErrorRegionHeartbeatResponse(server, s.clusterID, pdpb.ErrorType_UNKNOWN, msg)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
 		}
-		resp.Header = s.errorHeader(pberr)
 
-		return resp, nil
-	}
+		err = cluster.cachedCluster.handleRegionHeartbeat(region)
+		if err != nil {
+			msg := errors.Trace(err).Error()
+			err = sendErrorRegionHeartbeatResponse(server, s.clusterID, pdpb.ErrorType_UNKNOWN, msg)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
 
-	err := cluster.cachedCluster.handleRegionHeartbeat(region)
-	if err != nil {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: errors.Trace(err).Error(),
+		var resp *pdpb.RegionHeartbeatResponse
+		resp, err = cluster.handleRegionHeartbeat(region)
+		if err != nil {
+			msg := errors.Trace(err).Error()
+			err = sendErrorRegionHeartbeatResponse(server, s.clusterID, pdpb.ErrorType_UNKNOWN, msg)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		if resp == nil {
+			if s.cfg.regionHeartbeatUnaryMode {
+				// A workaround for passing tests, remove it ASAP.
```

Comment:
Yes, for passing tests, current tests still stay in unary model, one request one response, I will refactor those tests later.